tmk/cpp/bin/tmk-two-level-score.cpp

```diff
@@ -142,39 +150,66 @@ int main(int argc, char** argv) {
   // COMPUTE AND PRINT SCORES
-  for (const auto& it1 : metadataToFeatures) {
-    const std::string& metadata1 = it1.first;
-    std::shared_ptr<TMKFeatureVectors> pfv1 = it1.second;
-
-    for (const auto& it2 : metadataToFeatures) {
-      const std::string& metadata2 = it2.first;
-      std::shared_ptr<TMKFeatureVectors> pfv2 = it2.second;
-
-      if (!TMKFeatureVectors::areCompatible(*pfv1, *pfv2)) {
-        fprintf(
-            stderr,
-            "%s: immiscible provenances:\n%s\n%s\n",
-            argv[0],
-            metadata1.c_str(),
-            metadata2.c_str());
-        exit(1);
-      }
 
-      // Don't compare videos to themselves; don't do comparisons twice
-      // (A vs. B, then B vs. A).
-      bool skipThisPair =
-          includeSelf ? metadata1 > metadata2 : metadata1 >= metadata2;
-      if (skipThisPair) {
-        continue;
-      }
+  std::chrono::time_point<std::chrono::system_clock> startScores =
+      std::chrono::system_clock::now();
 
-      float s1 = TMKFeatureVectors::computeLevel1Score(*pfv1, *pfv2);
-      if (s1 >= c1) {
-        float s2 = TMKFeatureVectors::computeLevel2Score(*pfv1, *pfv2);
-        printf(
-            "%.6f %.6f %s %s\n", s1, s2, metadata1.c_str(), metadata2.c_str());
-      }
-    }
+  if (verbose) {
+    printf("\n");
+    printf("CALCULATING THE SCORES\n");
+  }
+
+  #pragma omp parallel
+  #pragma omp single nowait
+  {
+    for (const auto& it1 : metadataToFeatures) {
+      #pragma omp task firstprivate(it1)
+      {
+        const std::string& metadata1 = it1.first;
+        std::shared_ptr<TMKFeatureVectors> pfv1 = it1.second;
+
+        for (const auto& it2 : metadataToFeatures) {
+          #pragma omp task firstprivate(it2)
+          {
+            const std::string& metadata2 = it2.first;
+            std::shared_ptr<TMKFeatureVectors> pfv2 = it2.second;
+
+            if (!TMKFeatureVectors::areCompatible(*pfv1, *pfv2)) {
+              fprintf(
+                  stderr,
+                  "%s: immiscible provenances:\n%s\n%s\n",
+                  argv[0],
+                  metadata1.c_str(),
+                  metadata2.c_str());
+              exit(1);
+            }
+
+            // Don't compare videos to themselves; don't do comparisons twice
+            // (A vs. B, then B vs. A).
+            bool skipThisPair =
+                includeSelf ? metadata1 > metadata2 : metadata1 >= metadata2;
+
+            if (!skipThisPair) {
+              float s1 = TMKFeatureVectors::computeLevel1Score(*pfv1, *pfv2);
+              if (s1 >= c1) {
+                float s2 = TMKFeatureVectors::computeLevel2Score(*pfv1, *pfv2);
+                printf(
+                    "%.6f %.6f %s %s\n", s1, s2, metadata1.c_str(), metadata2.c_str());
+              }
+            }
+          } // end omp task it2
+        } // end for loop it2
+      } // end omp task it1
+    } // end for loop it1
+  } // end parallel
+  #pragma omp taskwait
```

Comment:
(Dumb question): If you do single without nowait, does that just generate this wait here?

Reply:
I wish I had a deeper understanding of this library to answer your question. I'm not familiar enough with it, and I'm confident there are some improvements that could be made! I tried a few different methods until I stumbled on this one that worked.

https://stackoverflow.com/a/13810539/4626770

Reply:
My level of expertise:

![image](https://user-images.githubusercontent.com/1654004/160634798-fcff8e01-de0f-47f0-ab95-f8b85f9ae213.png)